File: go/pattern.go
package nxs

// Access pattern detector (Adaptive-prefetch-spec §4).

const (
	SequentialThreshold        = 10
	RandomThreshold            = 100
	HistorySize                = 32
	MinObservations            = 8
	UpgradeSequentialThreshold = 100
)

// AccessPattern classifies observed record access.
type AccessPattern int

const (
	PatternUnknown AccessPattern = iota
	PatternSequential
	PatternRandom
	PatternMixed
)

func (p AccessPattern) String() string {
	switch p {
	case PatternSequential:
		return "sequential"
	case PatternRandom:
		return "random"
	case PatternMixed:
		return "mixed"
	default:
		return "unknown"
	}
}

// AccessPatternDetector observes record/seek indices and classifies access patterns.
type AccessPatternDetector struct {
	accesses       [HistorySize]int64
	writePos       int
	filled         int
	sequentialRuns uint32
	randomJumps    uint32
	lastIndex      int64
}

// NewAccessPatternDetector returns an empty detector.
func NewAccessPatternDetector() *AccessPatternDetector {
	d := &AccessPatternDetector{lastIndex: -1}
	for i := range d.accesses {
		d.accesses[i] = -1
	}
	return d
}

// SequentialRuns returns the count of sequential step observations.
func (d *AccessPatternDetector) SequentialRuns() uint32 {
	return d.sequentialRuns
}

// LastIndex returns the most recently observed record index, or -1 if none.
func (d *AccessPatternDetector) LastIndex() int64 {
	return d.lastIndex
}

// Observe records an access at index.
func (d *AccessPatternDetector) Observe(index int) {
	idx := int64(index)
	if d.lastIndex >= 0 {
		var delta uint64
		if idx >= d.lastIndex {
			delta = uint64(idx - d.lastIndex)
		} else {
			delta = uint64(d.lastIndex - idx)
		}
		if delta <= SequentialThreshold {
			if d.sequentialRuns < ^uint32(0) {
				d.sequentialRuns++
			}
		} else if delta > RandomThreshold {
			if d.randomJumps < ^uint32(0) {
				d.randomJumps++
			}
		}
	}
	d.accesses[d.writePos] = idx
	d.writePos = (d.writePos + 1) % HistorySize
	if d.filled < HistorySize {
		d.filled++
	}
	d.lastIndex = idx
}

// Pattern returns the current classification.
func (d *AccessPatternDetector) Pattern() AccessPattern {
	total := int(d.sequentialRuns) + int(d.randomJumps)
	if total < MinObservations {
		return PatternUnknown
	}
	if d.sequentialRuns > d.randomJumps*3 {
		return PatternSequential
	}
	if d.randomJumps > d.sequentialRuns {
		return PatternRandom
	}
	return PatternMixed
}

// PredictNext returns predicted next record indices when pattern is sequential (§4.4).
func (d *AccessPatternDetector) PredictNext(depth, recordCount int) []int {
	if d.Pattern() != PatternSequential || d.lastIndex < 0 {
		return nil
	}
	start := int(d.lastIndex) + 1
	out := make([]int, 0, depth)
	for i := 0; i < depth; i++ {
		idx := start + i
		if idx < recordCount {
			out = append(out, idx)
		}
	}
	return out
}

File: ruby/pattern.rb
# frozen_string_literal: true

# Access pattern detector (Adaptive-prefetch-spec §4).

module Nxs
  SEQUENTIAL_THRESHOLD = 10
  RANDOM_THRESHOLD = 100
  HISTORY_SIZE = 32
  MIN_OBSERVATIONS = 8
  UPGRADE_SEQUENTIAL_THRESHOLD = 100

  PATTERN_UNKNOWN = 'unknown'
  PATTERN_SEQUENTIAL = 'sequential'
  PATTERN_RANDOM = 'random'
  PATTERN_MIXED = 'mixed'

  # Observes record(index) / seek calls and classifies access patterns.
  class AccessPatternDetector
    def initialize
      @accesses = Array.new(HISTORY_SIZE, -1)
      @write_pos = 0
      @filled = 0
      @sequential_runs = 0
      @random_jumps = 0
      @last_index = -1
    end

    attr_reader :sequential_runs, :last_index

    def observe(index)
      idx = index
      if @last_index >= 0
        delta = (idx - @last_index).abs
        if delta <= SEQUENTIAL_THRESHOLD
          @sequential_runs = [@sequential_runs + 1, 0xffffffff].min
        elsif delta > RANDOM_THRESHOLD
          @random_jumps = [@random_jumps + 1, 0xffffffff].min
        end
      end
      @accesses[@write_pos] = idx
      @write_pos = (@write_pos + 1) % HISTORY_SIZE
      @filled += 1 if @filled < HISTORY_SIZE
      @last_index = idx
    end

    def pattern
      total = @sequential_runs + @random_jumps
      return PATTERN_UNKNOWN if total < MIN_OBSERVATIONS

      return PATTERN_SEQUENTIAL if @sequential_runs > @random_jumps * 3
      return PATTERN_RANDOM if @random_jumps > @sequential_runs

      PATTERN_MIXED
    end

    # Predicted next record indices when pattern is sequential (§4.4).
    def predict_next(depth, record_count)
      return [] unless pattern == PATTERN_SEQUENTIAL && @last_index >= 0

      start = @last_index + 1
      out = []
      depth.times do |i|
        idx = start + i
        out << idx if idx < record_count
      end
      out
    end
  end
end
